src/v/cluster/archival/segment_reupload.h

```diff
@@ -47,26 +78,36 @@ class segment_collector {
     /// \param max_uploaded_segment_size is a size limit for the offset range
     /// \param end_inclusive is a target for the end offset (if not set the
     ///        collector will try to match the size only)
+    /// \param end_exclusive last stable offset (unadjusted) of the target
+    ///        partition. in new upload mode, we won't upload a segment if
+    ///        it's dirty offset exceeds this value
+    /// \param flush_offset offset specified by a flush operation upstream,
+    ///        has no effect in reupload mode
     segment_collector(
       model::offset begin_inclusive,
       const cloud_storage::partition_manifest& manifest,
       const storage::log& log,
       size_t max_uploaded_segment_size,
-      std::optional<model::offset> end_inclusive = std::nullopt);
+      std::optional<model::offset> end_inclusive,
+      std::optional<model::offset> end_exclusive,
```

Comment:
if the 'end_exclusive' is an LSO does it have to be an optional?

Reply:
> does it have to be an optional

i wouldn't think so. i made it optional because the reupload paths don't seem to need it - or put another way, to avoid updating other constructor call sites. 

i don't have a strong opinion about it though, beyond a vague sense of keeping the diff small.